### PR TITLE
Allow server to return HTTP 4xx errors

### PR DESCRIPTION
**Pull request by duricanikolic**
#### What this PR does
This PR initializes `dskit`'s `server.Server` and `httpgrpc.Server` in such a way that they return errors with HTTP status code `4xx`, and therefore allow the latter to appear as `status_code` label in the request duration metrics.

#### Which issue(s) this PR fixes or relates to

Part of https://github.com/grafana/mimir/issues/1830

#### Checklist

- [x] Tests updated.
- [ ] Documentation added.
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`.
- [ ] [`about-versioning.md`](/docs/sources/mimir/configure/about-versioning.md) updated with experimental features.

### CODE REVIEW

**Review by dimitarvdimitrov — COMMENTED** [review]
Do I understand correctly that this change will also force the httpgrpc server to return an error and ignore the httpgrpc response when the code is 4xx?

There is this place in the query-fronted which needs an http response (taken from http grpc) in order to propagate the error back to the user. https://github.com/grafana/mimir/blob/9fd2b19a61acec59336d677a50270806dcc4482c/pkg/frontend/querymiddleware/codec.go#L406-L417

what's referred to as the `r` response in this snippet  should be what the querier computes here https://github.com/grafana/mimir/blob/5d0316a2e2175b9727aae115f4215107b6642ad7/pkg/querier/worker/scheduler_processor.go#L247

If my understanding is correct, then we will convert all 4xx promQL errors into 4xx errors. At the same time no tests are failing, so maybe I'm missing something

**Comment by pracucci** [comment]
> Do I understand correctly that this change will also force the httpgrpc server to return an error and ignore the httpgrpc response when the code is 4xx?

Yes, that's correct. I've done an analysis of possible side effects [here](https://github.com/grafana/mimir/issues/1830).

> There is this place in the query-fronted which needs an http response (taken from http grpc) in order to propagate the error back to the user.

_Going to take a look now._

**Comment by pracucci** [comment]
> what's referred to as the r response in this snippet should be what the querier computes here [mimir/pkg/querier/worker/scheduler_processor.go](https://github.com/grafana/mimir/blob/5d0316a2e2175b9727aae115f4215107b6642ad7/pkg/querier/worker/scheduler_processor.go#L247) Line 247 in [5d0316a](https://github.com/grafana/mimir/commit/5d0316a2e2175b9727aae115f4215107b6642ad7)
> `response, err := sp.handler.Handle(ctx, request)`

@dimitarvdimitrov That's correct, but I live it's handled correctly there. The `error` (which starting from this PR we'll get for 4xx errors too) is converted back to an HTTP response. When using httpgrpc, the HTTP response body is stored in `status.Details`, so `HTTPResponseFromError()` coverts it back to an HTTP response:

https://github.com/grafana/mimir/blob/fb7f57b663accfb3734f5c4851951965a5c3ce67/pkg/querier/worker/scheduler_processor.go#L247-L257

File: pkg/alertmanager/multitenant.go [hunk 1]
```diff
@@ -400,7 +400,7 @@ func createMultitenantAlertmanager(cfg *MultitenantAlertmanagerConfig, fallbackC
 		return nil, errors.Wrap(err, "failed to initialize Alertmanager's ring")
 	}
 
-	am.grpcServer = server.NewServer(&handlerForGRPCServer{am: am})
+	am.grpcServer = server.NewServer(&handlerForGRPCServer{am: am}, []server.Option{server.WithReturn4XXErrors}...)
```
[the remark this answers is not in the document]
Comment (duricanikolic) [reply pkg/alertmanager/multitenant.go:403]: You are right, I will change this. Thank you

**Comment by duricanikolic** [comment]
> is this PR a duplicate of #6983?

It is, and #6983 will be closed.

**Comment by dimitarvdimitrov** [comment]
> > Do I understand correctly that this change will also force the httpgrpc server to return an error and ignore the httpgrpc response when the code is 4xx?
> 
> Yes, that's correct. I've done an analysis of possible side effects [here](https://github.com/grafana/mimir/issues/1830).

I didn't check the linked issue before. The analysis there makes sense to me

**Review by dimitarvdimitrov — APPROVED** [review]
LGTM

File: pkg/ruler/remotequerier.go [hunk 2]
```diff
@@ -227,7 +229,9 @@ func (q *RemoteQuerier) query(ctx context.Context, query string, ts time.Time, l
 
 	resp, err := q.sendRequest(ctx, &req, logger)
 	if err != nil {
-		level.Warn(logger).Log("msg", "failed to remotely evaluate query expression", "err", err, "qs", query, "tm", ts)
+		if code := grpcutil.ErrorToStatusCode(err); code/100 == 5 {
```
Comment (pracucci) [inline pkg/ruler/remotequerier.go:232]: _Same here_

File: pkg/ruler/remotequerier_test.go [hunk 3]
```diff
@@ -709,27 +709,85 @@ func TestRemoteQuerier_BackoffRetry(t *testing.T) {
...
 			},
 			Body: []byte(`{
 				"status": "error","errorType": "execution"
 			}`),
-		}, nil
+		}
+		error4xx = status.Error(http.StatusUnprocessableEntity, "this is a 4xx error")
+		error5xx = status.Error(http.StatusInternalServerError, "this is a 5xx error")
+	)
+	testCases := map[string]struct {
```
Comment (pracucci) [inline pkg/ruler/remotequerier_test.go:725]: Add a test case returning a non httpgrpc error (e.g. `errors.New("mocked error")`). We expect the log.

**Review by pracucci — APPROVED** [review]
LGTM (modulo a couple of minor comments)

**Pull request merged**